File: Python/49_Group_Anagrams.py
from collections import defaultdict
from typing import List


class Solution:
    def groupAnagrams(selfself, strs: List[str]) -> List[List[str]]:
        # O(m * n)
        # m = num of words,
        # n = average char count of a word which is really about 6 or 7 if it's a real word
        anagram_groups = defaultdict(list)
        for word in strs:
            char_occurrences = [0] * 26 # a ... z
            for char in word:
                # gets number of letter in alphabet by subtracting ascii value of a
                char_occurrences[ord(char) - ord("a")] += 1

            anagram_groups[tuple(char_occurrences)].append(word)

        return list(anagram_groups.values())





solution = Solution()
print(solution.groupAnagrams(["eat","tea","tan","ate","nat","bat"]))

File: JavaScript/49_Group_Anagrams.js
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function(strs) {
    const anagrams = {}
    strs.forEach((item) => {
        const sorted = item.split('').sort().join('')
        if (anagrams[sorted]) {
            anagrams[sorted] = [...anagrams[sorted], item]
        } else {
            anagrams[sorted] = [item]
        }
        
    })
    // anagrams = {abt: [bat], ant: [nat, tan]}
    return Object.values(anagrams)
};